light-client/src/peer_list.rs

```diff
@@ -1,156 +1,206 @@
-    /// and mark it as faulty.
+    /// and mark it as faulty. Get a new witness from
+    /// the list of full nodes, if there are any left.
+    /// Returns the new witness, if any.
     ///
     /// ## Precondition
     /// - The given peer id must not be the primary peer id.
     /// - The given peer must be in the witness list
-    #[pre(peer_id != self.primary && self.witnesses.contains(&peer_id))]
+    #[pre(faulty_witness != self.primary && self.witnesses.contains(&faulty_witness))]
     #[post(Self::invariant(&self))]
-    pub fn mark_witness_as_faulty(&mut self, peer_id: PeerId) {
-        self.witnesses.remove(&peer_id);
-        self.faulty_nodes.insert(peer_id);
-    }
+    pub fn replace_faulty_witness(&mut self, faulty_witness: PeerId) -> Option<PeerId> {
+        let mut result = None;
 
-    /// Fill witness list with up to `n` full nodes
-    #[post(Self::invariant(&self))]
-    pub fn fill_witness_list(&mut self, n: usize) {
-        let new_witnesses: Vec<_> = self.full_nodes.iter().take(n).copied().collect();
-        for new_witness in new_witnesses {
-            self.full_nodes.remove(&new_witness);
+        self.witnesses.remove(&faulty_witness);
+
+        if let Some(new_witness) = self.full_nodes.iter().next().copied() {
             self.witnesses.insert(new_witness);
+            self.full_nodes.remove(&new_witness);
+            result = Some(new_witness);
         }
+
+        self.faulty_nodes.insert(faulty_witness);
+
+        result
     }
 
-    /// Swap the primary for the next available witness, if any.
+    /// Mark the primary as faulty and swap it for the next available witness, if any.
+    /// Returns the new primary on success.
     ///
     /// ## Errors
     /// - If there are no witness left, returns `ErrorKind::NoWitnessLeft`.
     #[post(ret.is_ok() ==> Self::invariant(&self))]
-    pub fn swap_primary(&mut self) -> Result<(), Error> {
-        while let Some(peer_id) = self.witnesses.iter().next() {
-            if peer_id != &self.primary {
-                self.primary = *peer_id;
-                return Ok(());
-            }
+    pub fn replace_faulty_primary(&mut self) -> Result<PeerId, Error> {
+        self.faulty_nodes.insert(self.primary);
+
+        if let Some(new_primary) = self.witnesses.iter().next().copied() {
+            self.primary = new_primary;
+            self.witnesses.remove(&new_primary);
+            return Ok(new_primary);
         }
 
         bail!(ErrorKind::NoWitnessLeft)
     }
 }
 
 /// A builder of `PeerList` with a fluent API.
-#[derive(Default)]
-pub struct PeerListBuilder {
-    instances: HashMap<PeerId, Instance>,
+pub struct PeerListBuilder<T> {
+    values: HashMap<PeerId, T>,
     primary: Option<PeerId>,
-    witnesses: HashSet<PeerId>,
-    full_nodes: HashSet<PeerId>,
-    faulty_nodes: HashSet<PeerId>,
+    witnesses: BTreeSet<PeerId>,
+    full_nodes: BTreeSet<PeerId>,
+    faulty_nodes: BTreeSet<PeerId>,
```

Comment:
It seems wasteful to see the exact same fields as in `PeerList`, why does the builder not hold on to a fresh `PeerList` and returning it on `build` instead of constructing it.

Reply:
I am more or less following the internal representation used by the [`derive_builder`](https://docs.rs/derive_builder/0.9.0/derive_builder/) crate, and didn't give it much more thought. But I agree that it's a bit wasteful and that what you suggest would work as well. I don't feel strongly either way.